[IMP] website_sale: faster recursive e-commerce category search

yajo opened the pull request:
https://github.com/odoo/odoo/pull/60834 has been rejected upstream as "fixed in v13". However I guess most of us still don't want sluggisy v12 e-commerces. Here it is for OCB.

---

When a website has a big category tree, without this patch it has to make a lot of queries just to build the tree. Now it's way faster because it solves all in 1 query.

For example, in a production database, and after warmup, this reduced the amount of queries from 369 to 229 while browsing a category. That's 140 less queries, a lot. Also, when browsing `/shop`, it reduced the queries from 326 to 172. The times inside and outside Postgres also improve according to my tests.

Note: this domain could be even more constrained to avoid reading all website categories if they are collapsed, but given https://github.com/odoo/odoo/pull/34249 is not in v12, it ends up producing lots of more queries (by using `child_of` or `parent_of` operators) and slowing more the result than just downloading all of them.

@Tecnativa TT26309

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

pedrobaeza commented on addons/website_sale/controllers/main.py:
```diff
@@ -230,10 +230,9 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
             post['attrib'] = attrib_list
 
-        Product = request.env['product.template'].with_context(bin_size=True)
+        search_categories = Product = request.env['product.template'].with_context(bin_size=True)
```
Why are you putting to `search_categories` the `bin_size=True` context?

yajo replied:
I guess it was just to save a LOC, but you're right; let me change that.

Commit:
[IMP] website_sale: faster recursive e-commerce category search

When a website has a big category tree, without this patch it has to make a lot of queries just to build the tree. Now it's way faster because it solves all in 1 query.

For example, in a production database, and after warmup, this reduced the amount of queries from 369 to 229 while browsing a category. That's 140 less queries, a lot. Also, when browsing `/shop`, it reduced the queries from 326 to 172. The times inside and outside Postgres also improve according to my tests.

Note: this domain could be even more constrained to avoid reading all website categories if they are collapsed, but given https://github.com/odoo/odoo/pull/34249 is not in v12, it ends up producing lots of more queries (by using `child_of` or `parent_of` operators) and slowing more the result than just downloading all of them.

@Tecnativa TT26309
